### ISSUE TITLE
[HTTP Foundation] Setting false as Request::isMethodSafe() first parameter should have been deprecated on 4.1

### ISSUE THREAD

**Issue by blitux** [symfony/symfony#31323]
**Symfony version(s) affected**: symfony/http-foundation v4.1+

**Description**  

On https://github.com/symfony/http-foundation/blob/master/Request.php#L1435-L1452

```php
/**
 * Checks whether or not the method is safe.
 *
 * @see https://tools.ietf.org/html/rfc7231#section-4.2.1
 *
 * @param bool $andCacheable Adds the additional condition that the method should be cacheable. True by default.
 *
 * @return bool
 */
public function isMethodSafe(/* $andCacheable = true */)
{
    if (!\func_num_args() || func_get_arg(0)) {
        // setting $andCacheable to false should be deprecated in 4.1
        throw new \BadMethodCallException('Checking only for cacheable HTTP methods with Symfony\Component\HttpFoundation\Request::isMethodSafe() is not supported.');
    }

    return \in_array($this->getMethod(), ['GET', 'HEAD', 'OPTIONS', 'TRACE']);
}
```
It actually throws an exception when no argument is passed. This should have been deprecated on 4.1. Source: https://github.com/symfony/http-foundation/commit/ba431a05b43e4afe8059444318310f47240715db

**How to reproduce**  
Create a `Symfony\Component\HttpFoundation\Request` instance and then call the method `isMethodSafe` on it with no arguments. 

**What is expected**
It should not throw an exception when calling the method without arguments, and raise an error with E_USER_DEPRECATED code when it's called with arguments.

**Possible Solution**  

```php 
    public function isMethodSafe()
    {
        if (\func_num_args() > 0) {
            @trigger_error('Passing arguments to Symfony\Component\HttpFoundation\Request::isMethodSafe() is not supported. To check if method is cacheable, use Symfony\Component\HttpFoundation\Request::isMethodCacheable() instead.', E_USER_DEPRECATED);
        }

        return \in_array($this->getMethod(), ['GET', 'HEAD', 'OPTIONS', 'TRACE']);
    }

```

**Comment by apfelbox** [comment]
If I read it correctly, calling it without parameter (or with `true` as parameter) is deprecated – as the default parameter can't be changed due to BC.

So the first step is to just pass the parameter explicitly: `->isMethodSafe(false)` and then maybe in the next version the parameter will be removed completely.

**Comment by stof** [comment]
Calling it without argument is indeed deprecated in 3.4. So there is no bug here.

**Comment by blitux** [comment]
Ok, I did some extra code archeology and analysis about this issue.

TL;DR: Setting false as the first argument of isMethodSafe should have been deprecated on 4.1 and the code updated, but it was not. I've updated the report to reflect these findings.

This was the commit where the deprecation was set: https://github.com/symfony/http-foundation/commit/5f7526b9b1be376eca4e8090f44313d1b17ce839

```php
    public function isMethodSafe(/* $andCacheable = true */)
    {
        if (!func_num_args() || func_get_arg(0)) {
            // This deprecation should be turned into a BadMethodCallException in 4.0 (without adding the argument in the signature)
            // then setting $andCacheable to false should be deprecated in 4.1
            @trigger_error('Checking only for cacheable HTTP methods with Symfony\Component\HttpFoundation\Request::isMethodSafe() is deprecated since version 3.2 and will throw an exception in 4.0. Disable checking only for cacheable methods by calling the method with `false` as first argument or use the Request::isMethodCacheable() instead.', E_USER_DEPRECATED);

            return in_array($this->getMethod(), array('GET', 'HEAD'));
        }

        return in_array($this->getMethod(), array('GET', 'HEAD', 'OPTIONS', 'TRACE'));
    }
```

Then, on https://github.com/symfony/http-foundation/commit/ba431a05b43e4afe8059444318310f47240715db the functionality to check for cacheable methods is removed and an exception is raised. We still must pass `false` for the argument here.

```php
    public function isMethodSafe(/* $andCacheable = true */)
    {
        if (!func_num_args() || func_get_arg(0)) {
            // setting $andCacheable to false should be deprecated in 4.1
            throw new \BadMethodCallException('Checking only for cacheable HTTP methods with Symfony\Component\HttpFoundation\Request::isMethodSafe() is not supported.', E_USER_DEPRECATED);
        }

        return in_array($this->getMethod(), array('GET', 'HEAD', 'OPTIONS', 'TRACE'));
    }
```

So, we are right now at 4.2.7 and setting the first argument as false is not deprecated nor the functionality updated. This method should not receive a useless argument. There was also a commit from Fabien that fixed some deprecation messages which added some clarity, but it never reached 4.1 it seems:

```
Checking only for cacheable HTTP methods with Symfony\Component\HttpFoundation\Request::isMethodSafe() is deprecated since Symfony 3.2 and will throw an exception in 4.0. Disable checking only for cacheable methods by calling the method with `false` as first argument or use the Request::isMethodCacheable() instead
```

So, I've updated the report to suggest deprecating setting `false` as the first argument of the method.

**Comment by nicolas-grekas** [comment]
Is anyone willing to move this forward? @blitux maybe?

**Comment by dFayet** [comment]
If nobody wants it, I can take it.
Just to be sure, does the fix must to be on the 4.1 branch?

**Comment by fabpot** [comment]
@dFayet 4.1 is not maintained  anymore, so 4.2 is the current lower branch that is still maintained.

**Comment by dFayet** [comment]
Actually we should maybe add a deprecation if we pass `false` to the method but is it correct to throw an exception if you pass no argument or `true` and a deprecation when `false` is given?

And in 5.0 should'nt all those checks be removed? As the argument was introduced to prevent a BC break in a 2.7 minor version?
